modules/shared/src/main/scala/org/tessellation/currency/dataApplication/package.scala

```diff
@@ -1,33 +1,150 @@
+    Encoder.instance {
+      case v: FeeTransaction => v.asJson
+      case v: DataUpdate     => e.apply(v)
+    }
+
+  implicit def decoder(implicit d: Decoder[DataUpdate]): Decoder[DataTransaction] =
+    (c: HCursor) => Decoder[FeeTransaction].apply(c).orElse(d.apply(c))
+
+  private def collectTransactions[A <: DataTransaction](dataTransactions: List[DataTransactions])(
+    pf: PartialFunction[DataTransaction, A]
+  ): List[Signed[A]] =
+    dataTransactions.flatMap { currentDataTransactions =>
+      currentDataTransactions.collect { dataTransaction =>
+        dataTransaction.value match {
+          case transaction if pf.isDefinedAt(transaction) =>
+            Signed(pf(transaction), dataTransaction.proofs)
+        }
+      }
+    }
+
+  def getDataUpdates(dataTransactions: List[DataTransactions]): List[Signed[DataUpdate]] =
+    collectTransactions(dataTransactions) {
+      case dataUpdate: DataUpdate => dataUpdate
+    }
+
+  def getFeeTransactions(dataTransactions: List[DataTransactions]): List[Signed[FeeTransaction]] =
+    collectTransactions(dataTransactions) {
+      case feeTransaction: FeeTransaction => feeTransaction
+    }
+
+  def getDataTransactionHashes[F[_]: Async: Hasher](
+    data: List[DataTransactions],
+    serializeDataUpdate: DataUpdate => F[Array[Byte]]
+  ): F[List[NonEmptyList[hash.Hash]]] =
+    data.traverse { dataTransactions =>
+      dataTransactions.toList.traverse { signedTransaction =>
+        signedTransaction.value match {
+          case dataUpdate: DataUpdate =>
+            Signed(dataUpdate, signedTransaction.proofs)
+              .toHashed(serializeDataUpdate)
+              .map(_.hash)
+          case feeTransaction: FeeTransaction =>
+            Signed(feeTransaction, signedTransaction.proofs)
+              .toHashed(serialize[F])
+              .map(_.hash)
+        }
+      }.map(NonEmptyList.fromListUnsafe)
+    }
+
+  def getFeeTransactionOfDataUpdate[F[_]: Async](
+    dataTransactions: DataTransactions,
+    dataUpdate: DataUpdate,
+    serializeDataUpdate: DataUpdate => F[Array[Byte]]
+  ): F[Option[Signed[FeeTransaction]]] = {
+    val feeTransactions = dataTransactions.collect {
+      case Signed(feeTransaction: FeeTransaction, proofs) => Signed(feeTransaction, proofs)
+    }
+
+    serializeDataUpdate(dataUpdate).map { serializedDataUpdate =>
+      feeTransactions.find { feeTransaction =>
+        Hash.fromBytes(serializedDataUpdate) === feeTransaction.value.dataUpdateRef
+      }
+    }
+  }
+}
+
+trait DataUpdate extends DataTransaction
+
+@derive(decoder, encoder, order, ordering, show)
+case class FeeTransaction(
+  source: Address,
+  destination: Address,
+  amount: Amount,
+  dataUpdateRef: Hash
+) extends DataTransaction
+
+object FeeTransaction {
+  def serialize[F[_]: Async](feeTransaction: FeeTransaction): F[Array[Byte]] =
+    Async[F].delay(JsonBinarySerializer.serialize(feeTransaction))
+
+  def decoderFeeTransactionRequest[F[_]: Async, D <: DataUpdate](
```

Comment:
`request` is part of infrastructure and should not be defined in the domain if possible. hence, I'd move it into the Routes (unless it would require redefining it many times)

Reply:
Where in the routes should I put this?

Reply:
next to the usage

Reply:
@marcinwadon The thing here is that this function is called on the requestDecoder lifecycle function, and also in some unit tests, so, if I put this in the routes file I can't access from other files